dbms/src/Storages/DeltaMerge/File/DMFileWithVectorIndexBlockInputStream.cpp

```diff
@@ -0,0 +1,487 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Storages/DeltaMerge/File/DMFileWithVectorIndexBlockInputStream.h>
+#include <Storages/DeltaMerge/File/VectorColumnFromIndexReader.h>
+#include <Storages/DeltaMerge/Index/VectorIndexCache.h>
+#include <Storages/DeltaMerge/Index/VectorSearchPerf.h>
+#include <Storages/DeltaMerge/ScanContext.h>
+#include <Storages/S3/FileCache.h>
+#include <Storages/S3/FileCachePerf.h>
+
+#include <algorithm>
+
+
+namespace DB::ErrorCodes
+{
+extern const int S3_ERROR;
+} // namespace DB::ErrorCodes
+
+namespace DB::DM
+{
+
+DMFileWithVectorIndexBlockInputStream::DMFileWithVectorIndexBlockInputStream(
+    const ANNQueryInfoPtr & ann_query_info_,
+    const DMFilePtr & dmfile_,
+    Block && header_layout_,
+    DMFileReader && reader_,
+    ColumnDefine && vec_cd_,
+    const FileProviderPtr & file_provider_,
+    const ReadLimiterPtr & read_limiter_,
+    const ScanContextPtr & scan_context_,
+    const VectorIndexCachePtr & vec_index_cache_,
+    const BitmapFilterView & valid_rows_,
+    const String & tracing_id)
+    : log(Logger::get(tracing_id))
+    , ann_query_info(ann_query_info_)
+    , dmfile(dmfile_)
+    , header_layout(std::move(header_layout_))
+    , reader(std::move(reader_))
+    , vec_cd(std::move(vec_cd_))
+    , file_provider(file_provider_)
+    , read_limiter(read_limiter_)
+    , scan_context(scan_context_)
+    , vec_index_cache(vec_index_cache_)
+    , valid_rows(valid_rows_)
+{
+    RUNTIME_CHECK(ann_query_info);
+    RUNTIME_CHECK(vec_cd.id == ann_query_info->column_id());
+    for (const auto & cd : reader.read_columns)
+    {
+        RUNTIME_CHECK(header_layout.has(cd.name), cd.name);
+        RUNTIME_CHECK(cd.id != vec_cd.id);
+    }
+    RUNTIME_CHECK(header_layout.has(vec_cd.name));
+    RUNTIME_CHECK(header_layout.columns() == reader.read_columns.size() + 1);
+
+    // Fill start_offset_to_pack_id
+    const auto & pack_stats = dmfile->getPackStats();
+    start_offset_to_pack_id.reserve(pack_stats.size());
+    UInt32 start_offset = 0;
+    for (size_t pack_id = 0, pack_id_max = pack_stats.size(); pack_id < pack_id_max; pack_id++)
```

Comment:
Why is pack_id_max here?

Reply:
change to `for (size_t pack_id = 0; pack_id < pack_stats.size(); ++pack_id)`